File: backend/app/schemas/hank_intake.py
"""Bounded extraction suggestions and explicit employee-reviewed filing contracts."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.document import DocumentType
from app.schemas.base import UTCModel

IntakeKind = Literal['purchase_order', 'vendor_quote', 'packing_slip', 'material_certificate', 'drawing', 'other']
IntakeStatus = Literal['queued', 'analyzing', 'awaiting_review', 'planned', 'completed', 'failed', 'cancelled']
FieldName = Literal[
    'document_number',
    'vendor_name',
    'customer_name',
    'po_number',
    'receipt_number',
    'packing_slip_number',
    'part_number',
    'work_order_number',
    'revision',
    'heat_number',
    'lot_number',
    'quantity',
    'unit_price',
    'total',
    'currency',
    'date',
    'due_date',
]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra='forbid')


class IntakeEvidence(StrictModel):
    page: int = Field(ge=1, le=25)
    excerpt: str = Field(min_length=1, max_length=400)


class IntakeField(StrictModel):
    name: FieldName
    value: str | None = Field(default=None, max_length=300)
    confidence: Literal['high', 'low', 'unknown'] = 'unknown'
    evidence: list[IntakeEvidence] = Field(default_factory=list, max_length=3)


class IntakeLine(StrictModel):
    description: str = Field(default='', max_length=500)
    part_number: str | None = Field(default=None, max_length=100)
    quantity: str | None = Field(default=None, max_length=100)
    unit_price: str | None = Field(default=None, max_length=100)
    lot_number: str | None = Field(default=None, max_length=100)
    heat_number: str | None = Field(default=None, max_length=100)
    confidence: Literal['high', 'low', 'unknown'] = 'unknown'
    evidence: list[IntakeEvidence] = Field(default_factory=list, max_length=3)


class IntakeExtraction(StrictModel):
    classification: IntakeKind
    confidence: Literal['high', 'low', 'unknown']
    summary: str = Field(max_length=1200)
    evidence: list[IntakeEvidence] = Field(default_factory=list, max_length=3)
    fields: list[IntakeField] = Field(default_factory=list, max_length=30)
    lines: list[IntakeLine] = Field(default_factory=list, max_length=50)
    warnings: list[str] = Field(default_factory=list, max_length=10)


class IntakeMatch(StrictModel):
    kind: Literal['part', 'work_order', 'vendor', 'purchase_order', 'receipt']
    id: int
    label: str
    href: str
    reason: str


class IntakeAnalysis(IntakeExtraction):
    matches: list[IntakeMatch] = Field(default_factory=list, max_length=25)
    has_duplicates: bool = Field(
        default=False, description='Matching content exists; private intake IDs remain hidden.'
    )
    duplicate_file_ids: list[int] = Field(default_factory=list, max_length=20)
    duplicate_document_ids: list[int] = Field(default_factory=list, max_length=20)


class IntakeCommand(StrictModel):
    expected_company_id: int = Field(gt=0)
    expected_version: int = Field(ge=1)


class IntakeReviewedField(StrictModel):
    name: FieldName
    value: str | None = Field(default=None, max_length=300)


class IntakePlanInput(StrictModel):
    filing_mode: Literal['draft', 'release_receipt_certificate'] = 'draft'
    title: str = Field(min_length=1, max_length=255)
    document_type: DocumentType
    revision: str = Field(default='A', min_length=1, max_length=20)
    description: str | None = Field(default=None, max_length=3000)
    part_id: int | None = Field(default=None, gt=0)
    work_order_id: int | None = Field(default=None, gt=0)
    vendor_id: int | None = Field(default=None, gt=0)
    purchase_order_id: int | None = Field(default=None, gt=0)
    receipt_id: int | None = Field(default=None, gt=0)
    reviewed_fields: list[IntakeReviewedField] = Field(default_factory=list, max_length=30)
    acknowledge_duplicate: bool = False


class IntakePlanCommand(IntakeCommand):
    plan: IntakePlanInput


class IntakePlanResponse(StrictModel):
    input: IntakePlanInput
    changes: list[str]
    warnings: list[str]
    references: list[IntakeMatch]


class IntakeReceipt(StrictModel):
    document_id: int
    document_number: str
    href: str
    summary: str
    references: list[IntakeMatch]
    warnings: list[str]


class IntakeFileResponse(UTCModel):
    id: int
    batch_id: int
    company_id: int
    filename: str
    file_size: int
    content_sha256: str
    page_count: int | None
    status: IntakeStatus
    version: int
    source_url: str
    analysis: IntakeAnalysis | None
    plan: IntakePlanResponse | None
    result: IntakeReceipt | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None


class IntakeBatchResponse(UTCModel):
    id: int
    company_id: int
    request_key: str
    created_at: datetime
    files: list[IntakeFileResponse]


class IntakeBatchList(StrictModel):
    batches: list[IntakeBatchResponse]
    has_more: bool
    next_before_id: int | None

File: frontend/src/types/hankIntake.ts
export type HankIntakeStatus =
  | 'queued'
  | 'analyzing'
  | 'awaiting_review'
  | 'planned'
  | 'completed'
  | 'failed'
  | 'cancelled';
export type HankIntakeFieldName =
  | 'document_number'
  | 'vendor_name'
  | 'customer_name'
  | 'po_number'
  | 'receipt_number'
  | 'packing_slip_number'
  | 'part_number'
  | 'work_order_number'
  | 'revision'
  | 'heat_number'
  | 'lot_number'
  | 'quantity'
  | 'unit_price'
  | 'total'
  | 'currency'
  | 'date'
  | 'due_date';
export type HankSourceFormat = 'pdf' | 'docx' | 'xlsx' | 'xls';
export interface HankSourcePreview {
  format: HankSourceFormat;
  units: string[];
  labels: string[];
  warnings: string[];
}
export interface HankIntakeEvidence {
  page: number;
  locator?: string | null;
  excerpt: string;
}
export interface HankIntakeMatch {
  kind: 'part' | 'work_order' | 'vendor' | 'purchase_order' | 'receipt';
  id: number;
  label: string;
  href: string;
  reason: string;
}
export interface HankIntakeAnalysis {
  source_format?: HankSourceFormat;
  source_labels?: string[];
  classification: 'purchase_order' | 'vendor_quote' | 'packing_slip' | 'material_certificate' | 'drawing' | 'other';
  confidence: 'high' | 'low' | 'unknown';
  summary: string;
  evidence: HankIntakeEvidence[];
  fields: Array<{
    name: HankIntakeFieldName;
    value: string | null;
    confidence: 'high' | 'low' | 'unknown';
    evidence: HankIntakeEvidence[];
  }>;
  lines: Array<{
    description: string;
    part_number: string | null;
    quantity: string | null;
    unit_price: string | null;
    unit_of_measure?: string | null;
    lot_number: string | null;
    heat_number: string | null;
    confidence: 'high' | 'low' | 'unknown';
    evidence: HankIntakeEvidence[];
  }>;
  warnings: string[];
  matches: HankIntakeMatch[];
  has_duplicates?: boolean;
  duplicate_file_ids: number[];
  duplicate_document_ids: number[];
}
export interface HankIntakePlan {
  filing_mode: 'draft' | 'release_receipt_certificate';
  title: string;
  document_type: string;
  revision: string;
  description?: string | null;
  part_id?: number | null;
  work_order_id?: number | null;
  vendor_id?: number | null;
  purchase_order_id?: number | null;
  receipt_id?: number | null;
  reviewed_fields: Array<{ name: HankIntakeFieldName; value: string | null }>;
  acknowledge_duplicate: boolean;
}
export interface HankIntakeFile {
  source_format?: HankSourceFormat;
  source_labels?: string[];
  id: number;
  batch_id: number;
  company_id: number;
  filename: string;
  file_size: number;
  content_sha256: string;
  page_count: number | null;
  status: HankIntakeStatus;
  version: number;
  source_url: string;
  analysis: HankIntakeAnalysis | null;
  plan: { input: HankIntakePlan; changes: string[]; warnings: string[]; references: HankIntakeMatch[] } | null;
  result: {
    document_id: number;
    document_number: string;
    href: string;
    summary: string;
    references: HankIntakeMatch[];
    warnings: string[];
  } | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
}
export interface HankIntakeBatch {
  id: number;
  company_id: number;
  request_key: string;
  created_at: string;
  files: HankIntakeFile[];
}
export interface HankIntakeBatchList {
  batches: HankIntakeBatch[];
  has_more: boolean;
  next_before_id: number | null;
}

/** Source evidence and deterministic PO matches; this read does not receive inventory. */
export interface HankIntakeReceivingDraft {
  file_id: number;
  file_version: number;
  company_id: number;
  filename: string;
  purchase_order_id: number | null;
  purchase_orders: Array<{ id: number; po_number: string; vendor_name: string; reason: string }>;
  packing_slip_number: string | null;
  lines: Array<{
    source_line_index: number;
    description: string;
    part_number: string | null;
    quantity: string | null;
    unit_of_measure: string | null;
    lot_number: string | null;
    heat_number: string | null;
    confidence: 'high' | 'low' | 'unknown';
    evidence: HankIntakeEvidence[];
    po_line_id: number | null;
    candidates: Array<{
      po_line_id: number;
      line_number: number;
      part_id: number;
      part_number: string;
      description: string;
      quantity_remaining: number;
      unit_of_measure: string;
    }>;
    quantity_received: number | null;
    warnings: string[];
  }>;
  warnings: string[];
  has_duplicates: boolean;
  requires_duplicate_acknowledgement: boolean;
}

/** Verified source values and scoped matches; importing is a separate reviewed task. */
export interface HankIntakePurchaseOrderDraft {
  file_id: number;
  file_version: number;
  company_id: number;
  filename: string;
  po_number: string | null;
  order_date: string | null;
  required_date: string | null;
  vendor_id: number | null;
  vendors: Array<{ id: number; code: string; name: string; reason: string }>;
  lines: Array<
    HankIntakeAnalysis['lines'][number] & {
      source_line_index: number;
      part_id: number | null;
      candidates: Array<{ id: number; part_number: string; name: string; unit_of_measure: string }>;
      quantity_ordered: number | null;
      unit_price_amount: number | null;
      warnings: string[];
    }
  >;
  warnings: string[];
  has_duplicates: boolean;
  blocked_reason: string | null;
  can_ready_for_receiving: boolean;
  existing_purchase_orders: Array<{ id: number; po_number: string; href: string }>;
}
